Fix @workflow when executing user has no permissions to access `review_history` in target state.

deiferni opened the pull request:
We have stumbled upon an edge-case when executing workflow transitions via the `@workflow` service endpoint when:
- the state variable `review_history` in the workflow is protected with a `guard-permission`
- executing a workflow-transition removes that permission for the executing user in the target state of the transition

An example workflow to illustrate the issue is provided in this PR. It has the following properties:
- two states `restricted` and `unrestricted` (default)
- two transitions `restrict` and `show` to change between the two states
- the `restricted` state removes the `View` permission for everyone but a `Manager` or a `Site Administrator`
- the variable `review_history` is protected with a `guard-permission` using the `View` permission

There is a test to illustrate the problem:
- when a user with none of the above roles executes the `restrict` transition from the `unrestricted` into the `restricted` state the transition is executed successfully, but
- retrieving the workflow history in the endpoint after the transition has been executed will fail in https://github.com/plone/plone.restapi/blob/f01ec6b8896a76d2877bb0ef3a29fb7964fac3f3/src/plone/restapi/services/workflow/transition.py#L80 with the following traceback
   ```
   ...
  File "plone.restapi/src/plone/restapi/services/workflow/transition.py", line 80, in reply
    history = self.wftool.getInfoFor(self.context, "review_history")
  File "eggs/Products.CMFCore-2.4.8-py3.8.egg/Products/CMFCore/WorkflowTool.py", line 288, in getInfoFor
    raise WorkflowException(msg)
  Products.CMFCore.WorkflowCore.WorkflowException: Could not get info: ${name}
   ```
- so the issue surfaces after successful execution but while attempting to serialize the JSON-response 

We provide a solution to this issue by accessing `review_history` with elevated privileges once the transition has been done using https://docs.plone.org/develop/plone/security/permissions.html#bypassing-permission-checks. This is safe because:
- only the last entry in the `review_history` is returned
- this is the workflow transition that the user just executed, they can know about it
- executing the transition itself is of course still done with normal privileges and will fail first if the user does not have the required permissions

Fixes https://github.com/plone/plone.restapi/issues/999

mister-roboto commented:
@deiferni thanks for creating this Pull Request and help improve Plone!

To ensure that these changes do not break other parts of Plone, the Plone test suite matrix needs to pass.

Whenever you feel that the pull request is ready to be tested, either start all jenkins jobs pull requests by yourself, or simply add a comment in this pull request stating:

```
@jenkins-plone-org please run jobs
```

With this simple comment all the jobs will be started automatically.

Happy hacking!

deiferni commented:
@jenkins-plone-org please run jobs

deiferni commented:
@jenkins-plone-org please run jobs

deiferni commented:
@jenkins-plone-org please run jobs

lukasgraf reviewed (approved):
LGTM 👍

jensens commented on src/plone/restapi/configure.zcml:
```diff
@@ -38,6 +38,15 @@
+      provides="Products.GenericSetup.interfaces.EXTENSION"
+      for="Products.CMFPlone.interfaces.ITestCasePloneSiteRoot"
+      />
+
```
the profile must be excluded in https://github.com/plone/plone.restapi/blob/master/src/plone/restapi/setuphandlers.py#L21

jensens replied:
@deiferni this need to be resolved before merge.

deiferni replied:
@jensens i have amended the desired changes in https://github.com/plone/plone.restapi/blob/aa99141eeaeb742aa7c1852d48d373c09d1a26b1/src/plone/restapi/setuphandlers.py#L21 yesterday, but i am currently fighting with what to me seem to be unrelated test-failures. will request re-review when it is ready!

deiferni commented:
@jenkins-plone-org please run jobs

deiferni commented:
@jenkins-plone-org please run jobs

deiferni commented:
@jenkins-plone-org please run jobs

deiferni commented:
@jensens thanks for your feedback! tests are now green, ready for re-review. i've amended my last commit so that the profile is excluded.

tisto commented:
Please do not merge for now. I have to clean up the mess in master / 6.x.x branch first and make a release before we can merge and release this.

ale-rt commented on src/plone/restapi/configure.zcml:
```diff
@@ -38,6 +38,15 @@
+      directory="profiles/testing-workflows"
+      description="Adds sample workflows for testing"
+      provides="Products.GenericSetup.interfaces.EXTENSION"
+      for="Products.CMFPlone.interfaces.ITestCasePloneSiteRoot"
```
IMHO if this profile is needed for the tests, it should be moved inside the test folder and the zcml should be included by the test setup.
In that way you will not even need to bother with the `getNonInstallableProfiles`

deiferni replied:
I have followed the existing convention of where the test profiles are put, there is already a `testing` profile in the same directory. In principle i agree but i'd rather not change too much for just this hotfix, so if you don't object i'd prefer to leave it where it is.

ale-rt commented on src/plone/restapi/services/workflow/transition.py:
```diff
@@ -77,20 +90,40 @@ def reply(self):
-                self.wftool.getTitleForStateOnType(
-                    action["review_state"], self.context.portal_type
-                ).decode("utf8")
+        sm = getSecurityManager()
```
I would use a context manager here

deiferni replied:
Ok, i have amended a context manager in my last commit and moved the whole security manager management into it.

deiferni commented:
@jenkins-plone-org please run jobs

deiferni commented:
@jenkins-plone-org please run jobs

deiferni commented:
@jenkins-plone-org please run jobs

deiferni commented:
@ale-rt thanks for your feedback, i have implemented or responded, in case you want to have another look. the tests are now also green again.

ale-rt reviewed (approved):
I am still of the idea that is a bad idea registering the testing profiles in the main `configure.zcml` and fiddle with `getNonInstallableProfiles` (see https://github.com/plone/plone.restapi/pull/1000/files#r530845175). Anyway it is ok for me to merge this PR in its current state!

ale-rt commented:
> Please do not merge for now. I have to clean up the mess in master / 6.x.x branch first and make a release before we can merge and release this.

@tisto can this be merged now?

deiferni commented:
@jenkins-plone-org please run jobs

deiferni commented:
Hey @tisto, I dropped the ball here for a while, but i'd like to pick this one up again.

- I have now rebased and made sure tests are green.
- Some things have changed in `black` so i have applied the necessary changes in a separate commit. I can also ship this in a separate PR if desired.

Ist there something still blocking this PR from your side? If not, please let me know what to do to move forward here. Thanks!

deiferni commented:
@jenkins-plone-org please run jobs

deiferni commented:
@tisto rebases after your black fixes in master and dropped local black fixes again. please let me know if anything else comes up and i will amend.

tisto commented:
@deiferni we can merge this now. I am just wondering about the testing profile in the main configure.zcml as @ale-rt mentioned previously. I don't have time right now to look into this right now. Though, won't the test profile show up in the available profiles to the user? Wasn't there a way to hide this?

deiferni commented:
@tisto we currently have the following strategy:
- the testing profile is defined in the same place as the other previously existing testing profile `plone.restapi:testing`
- it is then excluded in https://github.com/plone/plone.restapi/blob/04d3114ad25f5d0678405cb64fb8da4e2f5b2ac4/src/plone/restapi/setuphandlers.py#L21 like the other profiles, so it should not show up.

IIRC we agreed that moving all the testing profiles is out of scope for this bugfix. But i agree that having a separate `zcml` for testing only would by more elegant, so if you and @ale-rt want the profiles moved, i can do that first (separate PR probably). Let me know what you prefer!

deiferni commented:
thanks!
